src/worker/mod.rs

```diff
@@ -70,58 +99,80 @@ impl Worker {
             ))
+                as Box<
+                    dyn Poller<PollActivityTaskQueueResponse> + Send + Sync,
+                >)
         };
         let wf_task_poll_buffer = Box::new(WorkflowTaskPoller::new(
             wf_task_poll_buffer,
             sticky_queue_poller,
         ));
-        Self {
-            sticky_name: sticky_queue_name,
-            wf_task_poll_buffer,
-            at_task_mgr,
-            workflows_semaphore: Semaphore::new(config.max_outstanding_workflow_tasks),
+        Self::new_with_pollers(
             config,
-        }
+            sticky_queue_name,
+            sg,
+            wf_task_poll_buffer,
+            act_poll_buffer,
+        )
     }
 
-    #[cfg(test)]
     pub(crate) fn new_with_pollers(
         config: WorkerConfig,
         sticky_queue_name: Option<String>,
         sg: Arc<impl ServerGatewayApis + Send + Sync + 'static>,
         wft_poller: BoxedWFPoller,
         act_poller: Option<BoxedActPoller>,
     ) -> Self {
-        let max_acts = config.max_outstanding_activities;
+        let cache_policy = if config.max_cached_workflows == 0 {
+            WorkflowCachingPolicy::NonSticky
+        } else {
+            WorkflowCachingPolicy::Sticky {
+                max_cached_workflows: config.max_cached_workflows,
+            }
+        };
+        let (wau_tx, wau_rx) = unbounded_channel();
+        let (shut_tx, shut_rx) = watch::channel(false);
         Self {
+            server_gateway: sg.clone(),
             sticky_name: sticky_queue_name,
             wf_task_poll_buffer: wft_poller,
-            at_task_mgr: act_poller.map(move |ap| WorkerActivityTasks::new(max_acts, ap, sg)),
+            wft_manager: WorkflowTaskManager::new(wau_tx, cache_policy),
+            at_task_mgr: act_poller
+                .map(|ap| WorkerActivityTasks::new(config.max_outstanding_activities, ap, sg)),
             workflows_semaphore: Semaphore::new(config.max_outstanding_workflow_tasks),
             config,
+            shutdown_requested: shut_rx,
+            shutdown_sender: shut_tx,
+            workflow_activations_update: Mutex::new(wau_rx),
         }
     }
 
     /// Tell the worker to begin the shutdown process. Can be used before [shutdown_complete] if
     /// polling should be ceased before it is possible to consume the worker instance.
-    pub(crate) fn notify_shutdown(&self) {
+    pub(crate) async fn notify_shutdown(&self) {
+        let _ = self.shutdown_sender.send(true);
         self.wf_task_poll_buffer.notify_shutdown();
         if let Some(b) = self.at_task_mgr.as_ref() {
             b.notify_shutdown();
         }
+        loop {
+            // Wait until all outstanding workflow tasks have been completed before shutting down
+            if self.outstanding_workflow_tasks() != 0 {
+                let mut waulock = self.workflow_activations_update.lock().await;
+                let _ = waulock.recv().await;
```

Comment:
Waiting here doesn't update the workflow task semaphore, I'm assuming this is OK because we're shutting down.

Reply:
yep